### ISSUE TITLE
Add -lm (Was: undefined reference to symbol 'ceil@@GLIBC_2.2.5)

### ISSUE THREAD

**Issue by edi-bice** [tensorflow/tensorflow#2291]
### Environment info

Operating System:

epel-release-6-8.noarch
redhat-release-server-6Server-6.7.0.3.el6.x86_64

Installed version of CUDA and cuDNN: 

None

If installed from sources, provide the commit hash:

f8eb1d70a7ea7dc2cd5e1eddde389395f88a6be9
### Steps to reproduce
1. bazel clean
2. ./configure
3. bazel build -c opt //tensorflow/tools/pip_package:build_pip_package
### Logs or other output that would be helpful

(If logs are large, please upload as attachment).

ERROR: /home/ebice/tensorflow/google/protobuf/BUILD:272:1: Linking of rule '//google/protobuf:protoc' failed: gcc failed: error executing command /opt/rh/devtoolset-2/root/usr/bin/gcc -o bazel-out/host/bin/google/protobuf/protoc -no-canonical-prefixes -B/opt/rh/devtoolset-2/root/usr/bin -pass-exit-codes '-Wl,--build-id=md5' '-Wl,--hash-style=gnu' ... (remaining 11 argument(s) skipped): com.google.devtools.build.lib.shell.BadExitStatusException: Process exited with status 1.
/opt/rh/devtoolset-2/root/usr/bin/ld: /opt/rh/devtoolset-2/root/usr/lib/gcc/x86_64-redhat-linux/4.8.2/libstdc++_nonshared.a(hashtable_c++0x44.o): undefined reference to symbol 'ceil@@GLIBC_2.2.5'
/opt/rh/devtoolset-2/root/usr/bin/ld: note: 'ceil@@GLIBC_2.2.5' is defined in DSO /lib64/libm.so.6 so try adding it to the linker command line
/lib64/libm.so.6: could not read symbols: Invalid operation
collect2: error: ld returned 1 exit status
Target //tensorflow/tools/pip_package:build_pip_package failed to build

**Comment by edi-bice** [comment]
Edited google/protobuf/BUILD

changing 

LINK_OPTS = ["-lpthread"]

to

LINK_OPTS = ["-lpthread", "-lm"]

and the build goes on

**Comment by girving** [comment]
@martinwicke: Do you know of a reason we can't add `-lm` in all cases?  Would it bloat the binary?

**Comment by martinwicke** [comment]
@petewarden would be most sensitive to bloat. I'm always in favor of -lm.

**Comment by girving** [comment]
@edi-bice: Let us know if you want to send us a PR with `-lm`!

**Comment by edi-bice** [comment]
the overhead is high with that one - you guys go ahead if you don't mind

**Comment by alzhusan** [comment]
@edi-bice Where to find the `google/protobuf/BUILD` file?

**Comment by xuanchien** [comment]
@edi-bice 
Where is `google/protobuf/BUILD`? I am having this issue but I can not find this file to edit like you mentioned. I'm totally stuck now.

**Comment by xuanchien** [comment]
For those who have a similar issue. I had to modify the content of the file `bazel-tensorflow/external/protobuf/BUILD`, re-ran bazel build command  and it helped to resolve this issue in my case.

Before:
```
LINK_OPTS = select({
    ":android": [],
    "//conditions:default": ["-lpthread"],
})
```

New:
```
LINK_OPTS = select({
    ":android": [],
    "//conditions:default": ["-lpthread","-lrt","-lm"],
})
```

**Comment by zym1010** [comment]
I think it's easiest to simply append `-lm` on everything, by hacking third_party/gpus/crosstool/CROSSTOOL. Check <https://github.com/leelabcnbc/DevOps/blob/master/Docker/tensorflow/0.10.0/centos6/py35/gpu/CROSSTOOL.tpl#L59>, and my answer on #1426. Interestingly, this problem only appear when compiling with GPU...

**Comment by i3v** [comment]
I [had](https://github.com/tensorflow/tensorflow/issues/110#issuecomment-265431453) to combine suggestions from @zym1010 and @xuanchien to build tensorflow 0.12rc0 on CentOS6.5. I've used [CROSSTOOL.tpl#L59](https://github.com/leelabcnbc/DevOps/blob/master/Docker/tensorflow/0.10.0/centos6/py35/gpu/CROSSTOOL.tpl#L59), as a place to add necessary linker flags, but I had to add `-lrt` flag.

**Comment by zym1010** [comment]
@i3v Thanks for feedback! My methods still works for 0.11, but I'm not sure of 0.12, as the latest stable Bazel (which somehow I assumed is needed to compile TF 0.12) won't compile on CentOS.

**Comment by yliu120** [comment]
"won't compile on CentOS" is not true. Most likely is your environment var settings are not correct.

Please read my doc for building the latest github master branch on CentOS 6.7 with gcc 4.9
http://biophysics.med.jhmi.edu/~yliu120/tensorflow.html

**Comment by i3v** [comment]
@zym1010 I might be wrong, that `-lrt` is required. The thing is - I found a suggestion to add `-lrt` before I found a suggestion to add `-lm`. And after first tests it looked like `-lrt` is enough. I was not hitting any errors. However, shortly after that, I've noticed your suggestion to add `-lm`, so I've added it as well :) .

I'm not aware of any real cases, where both flags are needed. Neither I know any cases that would indicate that one flag is better than the other.

Concerning Bazel - I was able to build TF 0.12 with Bazel 0.4.0 so I doubt that most recent version is absolutely necessary. Also, the only issue I ran into, when trying to compile it [was](https://github.com/bazelbuild/bazel/issues/2177) associated with the number of threads Bazel bootstrapping script creates.
